LandSandBoat/server
# Core: Long-distance/Chunked pathing support

**_I affirm:_**

## What does this pull request do?

Closes https://github.com/LandSandBoat/server/issues/3522

Allows mobs to navigate a zone's entire length from a single, non-specialised call to our existing pathing functions - without tanking performance and without them resorting to wallhacking. For use in Besieged, Campaign, etc.

### Work

- [x] Fafnir can long-distance path the whole length of Bhaflau Thickets without resorting to wallhacking or getting trapped in the tunnel local minima.
- [x] Improve Tracy plumbing for more data: https://github.com/LandSandBoat/server/pull/6118
- [ ] Lots of testing on all types of mobs and path, and NPC pathing
- [ ] Abuse testing
- [ ] Enhance the debug logging in PathFind and NavMesh
- [ ] Performance profiling of different poly buffer sizes and overall load of chunking (if chunking even kicks in during regular use)
- [ ] Before/after performance profiling of regular use (The Kuftal Tunnel chase, etc.)
- [ ] New/before/after performance profiling of spawning 32 mobs, and sending them on long-distance pathing
- [ ] Audit the logic in PathFind
- [ ] Explain about the problem with local minima
- [ ] Explain about the problems with the logic in PathFind (when "wallhack" kicks in)
- [ ] Explain about "chunking" in Detour
- [ ] Explain about the cost of the poly buffer in NavMesh
- [ ] Explain about the final "bogus" point in an incomplete path
- [ ] Explain about incomplete paths
- [ ] Make the internals of pathfind and navmesh as const as possible
- [ ] Play with reducing poly search extents to as small as possible
- [x] Remove all allocations possible inside pathing routines
- [ ] Address all TODOs in code and PR
- [ ] Keep a list of waypoints, so you can manually define waypoints for a mob to navigate around/between
- [ ] Lots of mobs are trying to path from their valid start position to (0,0,0) and this causes a failure, what's going on here?
- [ ] The Debug logging for navmesh takes up a lot of time, how can I ask Tracy to ignore that specific line/symbol?
- [x] Move cpp members in navmesh into CNavmesh class, so they're not shared between instances (protecting against future threading issues)
- [ ] The logic inside FindPath should be: If I've requested a path and it's incomplete or invalid and the wallhack flag is on, that's when I should wallhack to my destination - to go after people who aren't on-mesh. If I have a complete or a partial path, then I should just follow that. This will need careful testing!

Notes:
```
My long-distance pathing work has ballooned out into changing the pathing->wallhacking logic

The current logic (if the wallhack flag is on):
I will try ask the navmesh a single time for for directions to my destination. I will put that destination as the last point of my directions so I will always wallhack there. If I got back no points, I will wallhack straight away.

The new logic:
I will ask the navmesh for directions. If I get back points I will follow them. If I reach the end of my points I will ask the navmesh again and follow them. Repeat. If I'm at my destination: success! If I asked the navmesh for more points, I got none back, and I'm not yet at my destination: I will wallhack the rest of the way to my destination.
 
It's also possible to look up why a mob might be wallhacking to go off-mesh
We can make the mob produce logs saying "I'm going off-mesh to chase after player X. Navmesh says that player X is n distance away from the nearest navmesh poly (x, y, z, zoneId)" 

That then produces logs that can be used to find broken/abuse points
```
<img width="827" alt="image" src="https://github.com/user-attachments/assets/dbff9102-1d1c-4e6f-b82f-0a0ab1e439bc">


```
1. Start Point
2. Local Minima Point
3. End Point
```

## Steps to test these changes

```
Info: Bhaflau Thickets
Start Point (Zone In): 383.7, -10.1, 10.0
End Point (North of the map): 70.5, -32.2, 643.5
```

- Truncate `mob_spawn_points` so the server is empty of anything that might try to path
- Enable `DEBUG_NAVMESH` in Logging
- `!zone {Bha` -> `{Bhaflau Thickets}`
- `!togglegm` (So you don't get aggro'd)
- `!fafnir`
- `!exec target:pathTo(70.5, -32.2, 643.5)`
- `/follow` Fafnir
- Observe Fafnir's path
- Observe the debug logging about the navmesh/pathing
